### DISCUSSION TITLE
Data right after the date [T060104X] TL3

### DISCUSSION THREAD

**Discussion by githubDante** [johanmeijer/grott#222] [General]
Does anyone have an idea what these 3 bytes contain ?!?

The data is extracted from result_string[136:148], in all templates (SPF, SPH) the record type starts @ 154, but there is data in the 3 skipped bytes
on TL3  these are always 020000 in the result_string 
`
b'\x02\x00\x00'
`
What this byte range (148:154) contain on SPF/SPH ?!?

**Comment by githubDante** [comment]
Further to this. The data behind the date (+ additional 2 bytes):

` 020000007c`

seems to be like a marker for end of the header and start of the actual data. This one is for the first group of registers. 
The marker for the second group seems to be the byte sequence:

` 007d00f9`

I see both in every packet type (060104X / 060103X / 050104X / 050103X). 
Can someone confirm my observations for other inverters SPH/SPF ?!? 
The packet in the example directory is shorter (265 bytes) and the first marker seems to be

`020000002c`

@johanmeijer from what inverter is this one?!?  Probably an inverter without second register group ?!?

**Comment by johanmeijer** [comment]
Ok lets see if we can find out what that means. 

As stated in PR #215:  
I would like to detect the inverter type automatically and I know it somewhere in the record (at least the registers being used are in the record) but do not know where and how to decode yet. That is why I introduced the recortype1 and 2 fields (only for debugging). I know that is not enough but it is a start.

Yes to to bytes after the 02 are the first register set (for mee with 2 TL-S inverters it is 02 0000 002C.  so it start with reg 0000 and ends at 002c (=45 regs in total 0-44). At pos 158 (the start of the record) + 4*45 (338)  the second set of registers start.  In my case 002D 0059. 

I also see spf records (eg layout: T060104SPF) with 03 at pos 148. So I agree with you this should be  the amount of register sets in the record. 

For the TL-S and SPF the reg set is different (see again #215 for the doc references). For TL-X (SPA), TL3-X (MAX, MID, MAC) , Storage(MIX), Storage(SPA), Storage (SPH) the registers are described in one document. Only the register sets being used are different as defined in the intro of the document (the 04 register sets are the one we see in the record): 

![image](https://user-images.githubusercontent.com/16508298/200675035-fd3c2c9a-c3ec-4c06-a338-81682e6ab004.png)

Until now I did not release that the TL-X was is this document (regs 3000-3124). So I used the TL-S as basis and did some debugging to match the record (the Tnnnnnnx records). 

I am going to look in the other records I have to see if we do not miss things, but for now this seems to be the solution.

**Comment by johanmeijer** [comment]
For the SPF record with 03 at 148 there are indeed 3 registers sets in the record: 

0000 002c
002d 0059
005a 0086

The strange thing here is that the record is not long enough to contain all register of the last set. 

The other things is that not all SPF inverters as a record layout with 3 regs. There are also ones with two sets. Unluckily  it are the same regs sets as defined for the TL-S but with total other meaning. So I do not think we can determine the inverter type from the data record (first quick conclusion).

**Comment by johanmeijer** [comment]
An other conclusion after testing some records. There are also SPH type inverters (SPH5000) that have defined 3 register sets in the record. This not documented in the document mentioned above. 

It also has registers 1125 - 1249 in it. 

We still are not complete, but more complete as before ;)

**Comment by githubDante** [comment]
The problem which I see are different inverter types which use the same set of registers. They should have something different. 
The difference between the MAX & MID series for example is:

 - packets from MAX have headers 06010[3-4] 
 - packets from MID have headers 05010[3-4]
 - also the header of the MID is shorter, the marker/register map is in the range [70:78], there is no empty space between the map and the date.

With all this information I will try to implement the detection in the doc tool.
If it works there it will work everywhere :smiley:

**Comment by githubDante** [comment]
Doc tool updated in #231
It should detect most types automatically. I don't have data from all inverter types therefore most of the tool is constructed around the info available here. Bugs are possible :)

**Reply by johanmeijer** [reply]
Oke but......

The difference of the 05/06 protocol is not being based on the the inverter type but on the software level of the datalogger. 

My datalogger (shinewifi-s) has started with sending t05xxxx records but after a firmware upgrade (somewhere in 2020) it changed in t06xxxx. 

So probably there are tl-s inverters out there sending t05 and others sending t06 records.

**Reply by githubDante** [reply]
OK, so a reliable dataset is needed in order to figure it out. In the case with the MID/MAX is not a big deal as they use the same registers, the MID just reports less data.

**Comment by githubDante** [comment]
@johanmeijer can you check your SPF records. Both which can be found in this repo are from buffered data, and for both the end of the header is marked with 3 to _03 0000 002C_ not _02 0000 002C_

**Comment by githubDante** [comment]
Found datarecord from MIN too -> type: 04  - detected and translated properly

**Comment by githubDante** [comment]
OK, I think I got it. SPF can contain 03 0000 002C or 02 0000 002C, the first means that there are 3 register ranges in the packet, the second one indicates that the packet contains only 2. If there are 3 markers the packet should be read to the end, just before the CRC check and it will contain data according to the range defined in the third marker. The rules for the next marker still apply (45*4 for SPF). As for the SPH with the 3 ranges in it, this probably is a SPA (the marker starts with 03?!?), which uses 9 registers in the range 1131:1140.

With all this information, why the inverter type is important (excluding the SPF)?!? The registers can be mapped by index:type and if the packet contain data at the respective position (after translation if needed) it will be shown, otherwise the value will be 0 or empty string. The users can decide which fields are not important to them and mark them.

**Comment by githubDante** [comment]
The doctool has been updated. If the inverter type is recognized it will create a map of the register markers in the packet. Then all registers as described in the documentation can be queried, and if the register is in the packet its value will be returned. The inverter detection is a bit tricky and not 100% accurate, but as long as the registers are recognized it should not be a problem.
If this or similar technique is integrated in the proxy/server versions of grott the layout files will not be needed. 
OK not quite as at least two similar files will still be needed, one for the inverters which use mapping as SPF (45 registers between the markers) and one for the rest with all registers with the respective description (human readable),index,type,divisor.

Thanks for the info shared here. It was really helpful. Thanks also to @egguy who provided the longer (839 bytes) packets with 3 register maps in them from SPH.
